src/client/signatures/canonical_json/src/ser.rs

```diff
@@ -0,0 +1,595 @@
+                let mut index = 0;
+
+                while index < 4 && string_iter.peek() != Some(&'}') && string_iter.peek() != None {
+
+                    match string_iter.peek() {
+                        Some(character) => characters.push(*character),
+                        None => break
+                    };
+
+                    string_iter.next();
+                    index += 1;
+                }
+
+                if string_iter.peek() == None {
+                    // could not find '}' bracket so must include '{' and following characters
+                    writer.write(&"{".as_bytes())?;
+                    writer.write(&characters.into_bytes())?;
+                } else if string_iter.peek() == Some(&'}') {
+                    // found '}' - remove '{' and '}' but must pad zeros
+                    if characters.len() == 0 {
+                        writer.write(&"{}".as_bytes())?;
+                    } else {
+                        writer.write(&std::iter::repeat("0").take(4 - characters.len()).collect::<String>().into_bytes())?;
+                        writer.write(&characters.into_bytes())?;
+                        string_iter.next(); // skip '}' 
+                    }
+                }
+            }
+
+            continue;
+        }
+
+        writer.write(curr_char.to_string().as_bytes())?;
+    }
+
+    Ok(())
+}
+
+/// Serialize a JSON value to String
+/// 
+/// # Examples
+/// ```rust
+/// # use canonical_json::ser::to_canonical_json;
+/// # use serde_json::json;
+/// # fn main() {
+///     to_canonical_json(&json!(null)); // returns "null"
+/// 
+///     to_canonical_json(&json!("test")); // returns "test"
+/// 
+///     to_canonical_json(&json!(10.0_f64.powf(21.0))); // returns "1e+21"
+/// 
+///     to_canonical_json(&json!({
+///         "a": "a",
+///         "id": "1",
+///         "b": "b"
+///     })); // returns "{"a":"a","b":"b","id":"1"}"; (orders object keys)
+/// 
+///     to_canonical_json(&json!(vec!["one", "two", "three"])); // returns "["one","two","three"]"
+/// # }
+/// 
+/// ```
+pub fn to_canonical_json(input: &serde_json::Value) -> String {
+
+    let string = vec![];
+    let null_string = "null".to_owned();
+    let mut serializer = serde_json::Serializer::with_formatter(string, JSONFormatter{});
+    match input.serialize(&mut serializer) {
+        Ok(()) => match String::from_utf8(serializer.into_inner()) {
+            Ok(serialized_string) => return serialized_string,
+            Err(error) => {
+                error!("Error converting bytes to string: {}", error);
+                return null_string;
+            }
+        },
+        Err(error) => {
+            error!("Error serializing JSON : {}", error);
+            return null_string;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use serde_json::json;
+    use log::{debug};
+    use env_logger;
+    use super::{to_canonical_json};
+
+    struct Test<'a> {
```

Comment:
Don't change it now, but for next time, there's still a lot of boilerplate here: a `Test` type (with a lifetime, no less!) and a description.  You've got the same number of lines as the functions.  Next time, drop the descriptions -- if you need, that's what comments are for.

The idiomatic way to do this is to use a Rust macro to be able to declare:

```rust
test!(input1, output1);
test!(input2, output2);
...
```
or similar.  And that's what I want: just 50 lines of pairs.  But now that it's here, don't bother -- you can spend a long time fighting Rust's macro system.